File: book/docs/slides/L23/estimate_pi.py
# estimate_pi.py
def estimate_pi(N: int) -> float:
    pi_forth = 0.0
    for n in range(N):
        pi_forth += (-1.0) ** n / (2.0 * n + 1.0)
    return pi_forth * 4


def main(argc: int, argv: list[str]) -> int:
    N = int(argv[1])
    print(f"\u03C0 = {estimate_pi(N)}")
    return 0


if __name__ == "__main__":
    import sys

    main(len(sys.argv), sys.argv)

File: book/docs/slides/L23/estimate_pi.cpp
// estimate_pi.cpp
#include <cmath>    // std::pow
#include <iomanip>  // std::setprecision
#include <iostream> // std::cout
#include <string>   // std::stoi

double estimate_pi(unsigned int N)
{
    double pi_fourth = 0.0;
    for (unsigned int i = 0; i < N; i++)
    {
        pi_fourth += std::pow(-1, i) * 1.0 / (2.0 * i + 1.0);
    }
    return 4.0 * pi_fourth;
}

int main(int argc, char **argv)
{

    unsigned N = std::stoi(argv[1]);
    std::cout << std::setprecision(20) << "\u03C0 = " << estimate_pi(N) << "\n";
    return 0;
}
